src/main/java/duke/Ui.java

```diff
@@ -0,0 +1,115 @@
+        Scanner sc = new Scanner(System.in);
+        Parser parser = new Parser();
+        taskList = storage.readFromFile();
+
+        while (true) {
+            String userInput = sc.nextLine();
+            if (userInput.equals("bye")) {
+                break;
+            } else if (userInput.equals("list")) {
+                display("Here are the tasks in your list:\n" + taskList.toString());
+            } else if (userInput.startsWith("mark")) {
+                // get index by splitting user input and get task at that index from list
+                int index = Integer.parseInt(userInput.split(" ")[1]);
+
+                if (index < 1 || index > taskList.getNumberOfTasks()) {
+                    throw new DukeIndexOutOfBoundsException("marked");
+                }
+
+                Task toBeMarked = taskList.getTaskAt(index - 1);
+                toBeMarked.mark();
+                display("Nice! I've marked this task as done:\n" + toBeMarked.toString());
+            } else if (userInput.startsWith("unmark")) {
+                // get index by splitting user input and get task at that index from list
+                int index = Integer.parseInt(userInput.split(" ")[1]);
+
+                if (index < 1 || index > taskList.getNumberOfTasks()) {
+                    throw new DukeIndexOutOfBoundsException("unmarked");
+                }
+
+                Task toBeUnmarked = taskList.getTaskAt(index - 1);
+                toBeUnmarked.unmark();
+                display("OK, I've marked this task as not done yet:\n" + toBeUnmarked.toString());
+            } else if (userInput.startsWith("delete")) {
+                int index = Integer.parseInt(userInput.split(" ")[1]);
+
+                if (index < 1 || index > taskList.getNumberOfTasks()) {
+                    throw new DukeIndexOutOfBoundsException("deleted");
+                }
+
+                Task toBeDeleted = taskList.getTaskAt(index - 1);
+                taskList.deleteTaskAt(index - 1);
+                display("Noted. I've removed this task:\n" + toBeDeleted.toString()
+                        + "\nNow you have " + taskList.getNumberOfTasks() + " tasks in the list.");
+            } else if (userInput.startsWith("find")) {
+                TaskList filtered = parser.getTaskList(userInput, taskList);
+                display("Here are the matching tasks in your list:\n" + filtered.toString());
+            } else {
+                Task add = parser.getTask(userInput);
+                try {
+                    taskList.addToList(add);
+                    display("Got it. I've added this task:\n" + add.toString()
+                            + "\nNow you have " + taskList.getNumberOfTasks() + " tasks in the list.");
+                } catch (NullPointerException e) {
+                    throw new DukeException("OOPS!!! Could not add task to the list");
+                }
+            }
+
+            storage.writeToFile(taskList);
+        }
```

Comment:
I like the very clear logical flow here, but is there a way you can shorten the code here by abstracting each 'case' using the enum class?